File: 백준/Gold/9024. 두 수의 합/두 수의 합.py
import sys
input=sys.stdin.readline

t=int(input())
for i in range(t):
    n,k=map(int,input().split())
    num=list(map(int,input().split()))
    num.sort()

    mini=200000000
    for j in range(n):
        l=j+1
        r= n-1
        while(l<=r):
            mid=(l+r)//2
            s=num[j]+num[mid]
            if(s>k):
                r=mid-1
            else:
                l=mid+1
            if(abs(k-s)<mini):
                cnt=1
                mini=abs(k-s)
            elif(abs(k-s)==mini):
                cnt+=1
    print(cnt)

File: 백준/Gold/9024. 두 수의 합/두 수의 합.js
const readline=require('readline')
const rl=readline.createInterface({
  input:process.stdin,
  output:process.stdout
})

// 두수의합? 

const input=[]
let iter=0
rl.on('line',(line)=>{
  input.push(line)
}).on('close',()=>{
  const t=Number(input[iter++])
  const answers=[]
  for(let _=0;_<t;_++){
    const [n,k]=input[iter++].split(' ').map(item=>Number(item))
    const l=input[iter++].split(' ').map(item=>Number(item))
    l.sort((a,b)=>a-b)

    let diff=[Math.abs(k-(l[0]+l[n-1])),0]
    let left=0
    let right=n-1
    while(left<right){
      const now=l[left]+l[right]
      const value=Math.abs(k-now)
      if (value===diff[0]){
        diff[1]+=1
      } else if (value<diff[0]){
        diff=[value,1]
      }
      if (k-now===0) {
        left+=1
        right-=1
      }
      else if (k-now>0) left+=1
      else right-=1
    }
    answers.push(diff[1])
  }
  console.log(answers.join('\n'))
  process.exit()
})

// -7, -4, -3, -2,  0, 1,  2,  5,  9, 12
// -7 12 -> 5 = 1
// -4 12 xxx
// -7 9 -> 2 = 2
